src/domain/core/src/services/push_status_service.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashMap;
+
+use internal_error::InternalError;
+use opendatafabric::{DatasetHandle, DatasetRefRemote};
+
+use crate::utils::metadata_chain_comparator::CompareChainsResult;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait PushStatusService: Send + Sync {
+    /// For a given dataset, returns list of push remotes with optional
+    /// synchronisation statuses. When could not fetch remote dataset info
+    /// to check state, status is None.
+    ///
+    /// Returns error when failed to determine head block of local dataset.
+    async fn get_statuses(
+        &self,
+        dataset_handle: &DatasetHandle,
+    ) -> Result<DatasetPushStatuses, InternalError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Clone, Debug)]
+pub struct DatasetPushStatuses {
+    pub statuses: HashMap<DatasetRefRemote, Option<CompareChainsResult>>,
```

Comment:
Weird to see `Option` here. When `None` status is returned?

I expected something more like `Result<CompareResult, CheckStatusError>`.

Reply:
I don't quite understand what constitutes an `unknown` status. The dataset is either in sync, not in sync, or some error happened while we were checking. If there was an error - I think we should surface it as such so that presentation layer could decide how to display it.

Reply:
Ok, I see that in service API you are already representing errors explicitly and there is no "unknown" state. In GQL API I think we should do the same. [Here](https://github.com/kamu-data/kamu-cli/blob/e496fe036b58367a5119a1c25721d8c333ec65af/src/adapter/graphql/src/mutations/dataset_metadata_mut.rs#L153) is a good example of how we represent logical errors as enums.

Reply:
done